core/domain/auth_domain.py

```diff
@@ -45,28 +46,149 @@ class AuthClaims(python_utils.OBJECT):
+                full user associated with that profile. None for full users.
+            deleted: bool. Whether the user has requested removal of their
+                account.
+        """
+        self.user_id = user_id
+        self.gae_id = gae_id
+        self.firebase_auth_id = firebase_auth_id
+        self.parent_user_id = parent_user_id
+        self.deleted = deleted
+
+    def validate(self):
+        """Checks that user_id, gae_id, firebase_auth_id, and parent_user_id
+        fields of this UserAuthDetails domain object are valid.
+
+        Raises:
+            ValidationError. The user_id is not specified.
+            ValidationError. The user_id is not a string.
+            ValidationError. The user_id has the wrong format.
+            ValidationError. The gae_id is not a string.
+            ValidationError. The firebase_auth_id is not a string.
+            ValidationError. The parent_user_id has the wrong format.
+            ValidationError. The parent_user_id is set for a full user.
+            ValidationError. The parent_user_id is not set for a profile user.
+        """
+        if not self.user_id:
+            raise utils.ValidationError('No user_id specified')
+
+        if not isinstance(self.user_id, python_utils.BASESTRING):
+            raise utils.ValidationError(
+                'user_id must be a string, but got %r' % self.user_id)
+
+        if not utils.is_user_id_valid(self.user_id):
+            raise utils.ValidationError(
+                'user_id=%r has the wrong format' % self.user_id)
+
+        if (self.gae_id is not None and
+                not isinstance(self.gae_id, python_utils.BASESTRING)):
+            raise utils.ValidationError(
+                'gae_id must be a string, but got %r' % self.gae_id)
+
+        if (self.firebase_auth_id is not None and
+                not isinstance(self.firebase_auth_id, python_utils.BASESTRING)):
+            raise utils.ValidationError(
+                'firebase_auth_id must be a string, but got %r' %
+                self.firebase_auth_id)
+
+        if (self.parent_user_id is not None and
+                not utils.is_user_id_valid(self.parent_user_id)):
+            raise utils.ValidationError(
+                'parent_user_id=%r has the wrong format' % self.parent_user_id)
+
+        if self.is_full_user() and self.parent_user_id:
+            raise utils.ValidationError(
+                'parent_user_id must not be set for a full user')
+
+        if not self.is_full_user() and not self.parent_user_id:
+            raise utils.ValidationError(
+                'parent_user_id must be set for a profile user')
+
+    def is_full_user(self):
+        """Returns whether the user is a full user (not a profile user).
+
+        Returns:
+            bool. Whether the user is a full user (not a profile user).
+        """
+        return self.gae_id is not None or self.firebase_auth_id is not None
+
+    def to_dict_with_changes(self):
```

Comment:
Just to check: conceptually, is it the case that this domain object represents a "user auth details" object, or the changes to such an object? These two concepts are different -- I've been treating it as the former while reviewing, but the explanation here suggests that it is representing the latter. Knowing which it is supposed to be will help me make further suggestions.

Reply:
The domain object represents the former, this method is the API to treat it as the latter.

Reply:
Ah, that's a problem actually. An object and the changes to it are not the same thing. For example, notice how we have a list of ExplorationChange objects when making an update to explorations, and these are different from the Exploration object itself.

Similarly, if we want to have an object with the conceptual meaning of "this represents changes to a user auth details object", that is OK, but let's actually define a new object and call it something like UserAuthChanges instead.

Reply:
I've done a lot of refactoring, PTAL

Reply:
Obsolete.